pkg/controller/pxcrestore/controller.go

```diff
@@ -105,370 +103,279 @@ func (r *ReconcilePerconaXtraDBClusterRestore) Reconcile(ctx context.Context, re
 	}
-	err = r.stopCluster(cluster.DeepCopy())
-	if err != nil {
-		err = errors.Wrapf(err, "stop cluster %s", cluster.Name)
-		return rr, err
+
+	if err := cluster.CheckNSetDefaults(r.serverVersion, log); err != nil {
+		statusState = api.RestoreFailed
+		statusMsg = err.Error()
+		return rr, errors.Wrap(err, "wrong PXC options")
 	}
 
-	log.Info("starting restore", "cluster", cr.Spec.PXCCluster, "backup", cr.Spec.BackupName)
-	err = r.setStatus(cr, api.RestoreRestore, "")
+	err = backup.CheckPITRErrors(ctx, r.client, r.clientcmd, cluster)
 	if err != nil {
-		err = errors.Wrap(err, "set status")
+		statusState = api.RestoreFailed
+		statusMsg = err.Error()
 		return rr, err
 	}
 
-	err = r.restore(ctx, cr, bcp, cluster)
+	bcp, err := getBackup(ctx, r.client, cr)
 	if err != nil {
-		err = errors.Wrap(err, "run restore")
-		return rr, err
+		statusState = api.RestoreFailed
+		statusMsg = err.Error()
+		return rr, errors.Wrap(err, "get backup")
 	}
 
-	log.Info("starting cluster", "cluster", cr.Spec.PXCCluster)
-	err = r.setStatus(cr, api.RestoreStartCluster, "")
+	restorer, err := r.getRestorer(cr, bcp, cluster)
 	if err != nil {
-		err = errors.Wrap(err, "set status")
-		return rr, err
+		return rr, errors.Wrap(err, "failed to get restorer")
 	}
 
-	if cr.Spec.PITR != nil {
-		oldSize := cluster.Spec.PXC.Size
-		oldUnsafe := cluster.Spec.AllowUnsafeConfig
-		cluster.Spec.PXC.Size = 1
-		cluster.Spec.AllowUnsafeConfig = true
-
-		if err := r.startCluster(cluster); err != nil {
-			return rr, errors.Wrap(err, "restart cluster for pitr")
+	switch statusState {
+	case api.RestoreNew:
+		annotations := cr.GetAnnotations()
+		_, unsafePITR := annotations[api.AnnotationUnsafePITR]
+		cond := meta.FindStatusCondition(bcp.Status.Conditions, api.BackupConditionPITRReady)
+		if cond != nil && cond.Status == metav1.ConditionFalse && !unsafePITR {
+			statusState = api.RestoreFailed
+			statusMsg = fmt.Sprintf("Backup doesn't guarantee consistent recovery with PITR. Annotate PerconaXtraDBClusterRestore with %s to force it.", api.AnnotationUnsafePITR)
+			return rr, nil
 		}
-
-		log.Info("point-in-time recovering", "cluster", cr.Spec.PXCCluster)
-		err = r.setStatus(cr, api.RestorePITR, "")
+		err = validate(ctx, restorer, cr)
 		if err != nil {
-			return rr, errors.Wrap(err, "set status")
+			if errors.Is(err, errWaitValidate) {
+				return rr, nil
+			}
+			err = errors.Wrap(err, "failed to validate restore job")
+			return rr, err
 		}
-
-		err = r.pitr(ctx, cr, bcp, cluster)
+		cr.Status.PXCSize = cluster.Spec.PXC.Size
+		cr.Status.AllowUnsafeConfig = cluster.Spec.AllowUnsafeConfig
+		log.Info("stopping cluster", "cluster", cr.Spec.PXCCluster)
+		statusState = api.RestoreStopCluster
+	case api.RestoreStopCluster:
+		err = stopCluster(ctx, r.client, cluster.DeepCopy())
 		if err != nil {
-			return rr, errors.Wrap(err, "run pitr")
+			switch err {
+			case errWaitingPods, errWaitingPVC:
```

Comment:
Could we check these errors on the `if` line so we can avoid this inner switch?

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/1610/commits/7b5a420742ce91caf7574135e02a02c633104621